core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -64,12 +64,79 @@ object RequestChannel extends Logging {
     }
   }
 
-  class Request(val processor: Int,
-                val context: RequestContext,
-                val startTimeNanos: Long,
+  object Request {
+    def apply(processor: Int,
+              context: RequestContext,
+              startTimeNanos: Long,
+              memoryPool: MemoryPool,
+              buffer: ByteBuffer,
+              metrics: RequestChannel.Metrics): Request = {
+      val session = Session(context.principal, context.clientAddress)
+      lazy val bodyAndSize: RequestAndSize = context.parseRequest(buffer)
+      lazy val summary: RequestSummary = new RequestSummary(
+        processor,
+        metrics,
+        context,
+        session,
+        startTimeNanos,
+        description = s"${context.header} -- ${bodyAndSize.request.toString(false)}",
+        detailedDescription = s"${context.header} -- ${bodyAndSize.request.toString(true)}",
```

Comment:
I don't think this achieves the goal because it retains a reference to the buffer. I need to take a closer look to figure out how we can be both lazy and avoid unnecessary retention.

Reply:
Thanks for the response @ijuma .  My understanding is that a reference to the buffer will remain in `RequestChannel.Request`, but it is released after its handled by `KafkaApis` by this line in the `KafkaRequestHandler`.

https://github.com/apache/kafka/blob/1.0.0/core/src/main/scala/kafka/server/KafkaRequestHandler.scala#L65

If this is true then the reference to the `buffer` is released ASAP.  The reference to the parsed `AbstractRequest` is held within `bodyAndSize` in `RequestChannel.Request`, which is no longer referenced in the response and has the potential to be deallocated by the GC sooner than it did before.

Reply:
I'm referring to the fact that `RequestSummary` is holding a reference to the buffer due to the lazy vals that are referenced from the by name parameters. The whole reason for the PR is for `RequestSummary` not to hold a reference to the buffer.

Reply:
I got it, you're referring `description`, `detailedDescription`, `requestBodyAndSize`, and `isFromFollower`.  That's why originally I wasn't using by name params to create `RequestSummary`, but it was probably being generated sooner than it need to be.

Basically we need the `RequestSummary` to be generated right before the request is handled in `KafkaApis` while also only parsing the `AbstractRequest` once.

I'll think on this too and follow up.

Reply:
Right. So, maybe the answer is to create the summary when creating `Response`

Reply:
That's similar to my original implementation.  I think the only difference was that `RequestSummary`  wasn't lazy and was being created as soon as the `RequestChannel.Request` was instantiated.

Reply:
Right, the issue with the original implementation is that `RequestSummary` was part of `Request`, but the separation was a bit muddled. So I thought it would be good to move it out of the `Request`. The original attempt was to do it before the `Request` is created whereas now we are thinking of doing it when we no longer need `Request` (i.e. when `Response` is created). Hopefully that sticks.

Reply:
I got it.  I'll follow up shortly.